Add option to log json using stdout adapter

indrekj opened the pull request:
STDOUT JSON logging allows us to transport logs to elasticsearch with context using STDOUT.

This PR also improves the logging performance.

Review commit-by-commit.

INF-1146

indrekj committed:
Update dependencies

indrekj committed:
Update author to techmovers

indrekj committed:
Pass adapter options to stdout

Now you can pass `json: true` to enable JSON STDOUT logs.

INF-1146

indrekj committed:
Remove winston adapter

From now on we are going to use only STDOUT adapter with regular and
JSON output.

indrekj committed:
Remove async, cycle and q

All these are unused dependencies

indrekj committed:
Remove default whitelist

These are very specific whitelist items. E.g. message might include a
chat message which must not be whitelisted. It makes more sense to let
the library user to define all the whitelist items.

indrekj committed:
Remove underscore dependency

V8 has many of these functions already builtin. Rather use plain JS when
possible.

indrekj committed:
Remove `should` package

We are already using chai. No need to mix these.

indrekj committed:
Remove duplicate assignment

indrekj committed:
Remove blacklist preprocessor

So far there has been no need for blacklist.

indrekj committed:
Use fixed mask value for excluded keys

There are sometimes values that are very long which we want to mask.
There's no reason to display 100 askerisks if 5 is also good enough.

This should also improve security a bit because now you don't even know
the value length.

Similar to
https://github.com/salemove/logasm/commit/ccaa16709dbd8ac96e7e678c662b7dd19f40089f

indrekj committed:
Prepare pointer regular expressions before using it

This greatly improves the whitelisting performance.

indrekj committed:
Remove _decodePointer from whitelist module

Users can replace ~0 with ~ themselves. We don't want to waste resources
for that here.

indrekj committed:
Replace regex matching with simple object lookup

This is a lot more effecient way to find whitelist matches. We do not
have to loop over all fields and use regex, instead we can use a simple
hashmap lookup.

This commit also removed matches for fields with tildes in it. They are
exceptions and we haven't seen them being used.

indrekj committed:
Clean whitelist module arguments order

Removed processData and made all functions to consistently use (value,
path) pairs.

indrekj committed:
Always whitelist /message field

04511e83dc3827478f02abb6bb6a8aa568091790 removed default whitelist. We
still need to ensure that /message is always whitelisted as this is
always logged.

indrekj committed:
Bump version to 1.0.0

indrekj commented on lib/logasm/preprocessors/whitelist.coffee:
```diff
@@ -1,15 +1,26 @@
 {isObject} = require('../object_helpers')
 
+isWildcardPointer = (pointer) ->
```
this flow is later replaced by a hashmap lookup https://github.com/salemove/node-logasm/pull/7/commits/6ea8b81044d6c83ce534a73c4381d3854ce71431 . probably no need to dive deep into this commit.

deiwin replied:
Thanks :D

ostap0207 commented:
Did you do or can do any performance testing? Could you put some results here?

indrekj commented:
Yes, I did some general performance testing. I used stdout adapter to print out 10K messages using about ~5 wildcard and 5 regular filters. Without wildcards it took 600ms, with wildcards (using regex) it took ~1.3 seconds. After switching to hashmap lookup both took ~600ms.

ostap0207 commented on lib/logasm/preprocessors/whitelist.coffee:
```diff
@@ -29,11 +29,6 @@ class Whitelist
     if pointer.charAt(pointer.length - 1) == '/'
       throw Error('Pointer should not contain trailing slash')
 
-  _decodePointer: (pointer) ->
```
Should update doc https://github.com/salemove/node-logasm#whitelist, saying that it's a subset of JSONPointer and `~` and `/` are not supported.

ostap0207 commented:
One comment, otherwise 👍 . One concern though about deleting functionality, such as blacklist and `~`. I guess it OK in this case as it looks no one else is really using this lib, but in other case it wouldn't seem like a right approach. Also if we only care here about what we use or don't use maybe we should make this repo private.

indrekj committed:
Reintroduce ~0 and ~1 encoded whitelist variables

From https://tools.ietf.org/html/rfc6901.

indrekj commented:
That's why I bumped the major version. @ostap0207 added `~` and `/` encoding back, was simple enough actually: https://github.com/salemove/node-logasm/pull/7/commits/76db97155a6201eb4b2143f1e496b5a4c2a2258b

deiwin commented on package.json:
```diff
@@ -1,7 +1,7 @@
   "name": "logasm",
   "version": "0.4.0",
-  "author": "Salemove <support@salemove.com>",
+  "author": "Salemove TechMovers <techmovers@salemove.com>",
```
SaleMove also has a capital M :)

deiwin commented on lib/logasm/adapters/stdout_adapter.coffee:
```diff
@@ -1,23 +1,24 @@
-      timestamp: ->
-        "#{new Date().toISOString()} ##{process.pid}"
+      json: Boolean(options.json)
+      timestamp: !Boolean(options.json) # Json adds @timestamp manually
```
Shouldn't it be the function it was previously when `options.json` is `false` (or falsy).

deiwin replied:
I guess you're doing something of the sort in `buildEvent`, but why? Also there `process.pid` part isn't there anymore. Was it useful?

indrekj replied:
the default timestamp is okay. Process pid is not useful at all with kubernetes (all processes are pid 0). I would say it wasn't even useful with the previous deployment.

deiwin commented on lib/logasm/adapters/stdout_adapter.coffee:
```diff
@@ -1,23 +1,24 @@
+      timestamp: !Boolean(options.json) # Json adds @timestamp manually
       level: level
       colorize: true
+      stringify: (obj) -> JSON.stringify(obj)
```
Kind of confused. Why is this here regardless of the `json` option?

indrekj replied:
`json: true` has `JSON.stringify(obj, null, 2)` which means it creates multiline json objects. We need one-line logs.

deiwin commented on lib/logasm/object_helpers.coffee:
```diff
@@ -0,0 +1,5 @@
+isObject = (obj) ->
+  type = typeof obj
+  type == 'function' || type == 'object' && Boolean(obj)
```
Isn't the underscore function `obj instanceof Object` instead?

indrekj replied:
nope https://github.com/jashkenas/underscore/blob/master/underscore.js#L1313-L1315

deiwin reviewed (commented):
> Yes, I did some general performance testing. I used stdout adapter to print out 10K messages using about ~5 wildcard and 5 regular filters. Without wildcards it took ~600ms, with wildcards (using regex which was preconstructed in constructor) it took ~1.3 seconds. After switching to hashmap lookup both took ~600ms.

Can you include a similar performance test in the project so that this would be automatically run for all PRs?

indrekj committed:
Add whitelist performance test

We want to keep our logging system as fast as possible. Otherwise it
will start interfering with the application code.

indrekj commented:
@deiwin Added https://github.com/salemove/node-logasm/pull/7/commits/f8940606817bf74628dcb74ad167e9ac009f91cb

Same test on master: `AssertionError: expected 3670 to be at most 100`.

indrekj commented:
!merge
